### DISCUSSION TITLE
How to pass a readable and writeable bytearray to a C module? Or how can I prevent the C module allocate memory section that is already used by micropython?

### DISCUSSION THREAD

**Discussion by nspsck** [micropython/micropython#12029] [ESP32] [answered]
Hi guys, I am currently working on a tft display driver. The tft panel is huge, 536*240. For 16bits-color, a whole frame will take roughly 250KB of ram. By doing this, I learned, that using a framebuffer in C modules can be very dangerous. Since the C modules have no idea about if either a section of RAM has been used or not, it could cause data corruption. In my case, the device just triggered a reset. Half of the amount RAM I needed is quite stable, but again, I have no idea if it is guaranteed, not to use the section that has meaningful data. 

If I use the bytearray, I could not find a way to modify it in the C module, hence I have to use a Frambuffer in the framebuf module. This cost me 22% performance. (Simply filling the whole screen with one color dropped down to 54 fps from 66 fps) So... ye... I really do not want to do complicated drawing in python, since they will be done very slowly. 

Since I am also new to C and python and tft drivers, please do assume that I haven't thought through some very simple and effective solutions and please do let me know about those! Thank you very much in advance!

ps. chunking is not preferred, unless there is a way to guarantee that the buffer for the chunks are safe to use.

**Comment by glenn20** [comment, marked as answer]
You can pass a readable and writable bytearray to a C module quite easily, in fact the following example function in a C module will accept a bytearray, memoryview or array (or any r/w python object which supports the "buffer protocol") (I'm not sure if that actually will help with your application):

```c
STATIC mp_obj_t my_function(mp_obj_t buf_obj) {
    // Get a pointer to the data of the supplied buffer (eg. bytearray)
    mp_buffer_info_t buf;
    mp_get_buffer_raise(buf_obj, &buf, MP_BUFFER_RW);

    // buf.len is length of the supplied data buffer.
    // buf.buf is a void * ptr to the contiguous data underlying the bytearray or array
    // Process the bytearray underlying data....

    return mp_const_none;
}
STATIC MP_DEFINE_CONST_FUN_OBJ_1(mymod_my_function, my_function);
```
I assume you know how to add that function to your module.

**Reply by glenn20** [reply]
Oh, and then you would call it with:

```py
import my_module

buffer = bytearray(size)
my_module.my_function(buffer)
```

**Reply by nspsck** [reply]
Hi, thank you for the help! The MP_BUFFER_RW is actually what i needed! Now here is another problem... The performance...

I have initialized it like such:
```C
// in the struct: 
typedef struct _mp_lcd_rm67162_obj_t {
    ...
    mp_buffer_info_t frame_buffer;
    ...
} mp_lcd_rm67162_obj_t;

// in the constuctor
mp_get_buffer_raise(args[ARG_buf].u_obj, &self->frame_buffer, MP_BUFFER_RW);
```
and later filled this using:
```C
STATIC void fill_color_buffer(mp_lcd_rm67162_obj_t *self, uint16_t color, int len /*in pixel*/) {
    uint16_t *buffer = self->frame_buffer.buf;
    for (int i = 0; i < len; i++) {
        *buffer++ = color;
    }
    write_color(self, self->frame_buffer.buf, len * 2);
}
```
However, the fps dropped down to 45 from 66, this is even slower than filling the buf with the framebuf module then later pass to the function (which is 54). This leads me into thinking that the buffer contains some slow PSRAM or RAM section which can not be accessed by the DMA... (How can I prevent this to happen?) Or the C module has to talk to the python intepreter... Also, drawing lines became inconsistent in performance... Or my filling methode is just slower and I do not know how to interact with this buf... Actually, this is more likely since the framebuf can perform better with the same allocated buf using: 
```python 
buf = bytearray(526*240*2)
```
I've tried to search for the framebuf module and found  a interesting function called "mp_obj_malloc()" in this file: [https://github.com/micropython/micropython/blob/master/extmod/modframebuf.c](https://github.com/micropython/micropython/blob/master/extmod/modframebuf.c). So, could this be the solution for my second approch? Could you please link me a documentation to all the mp_ functions, that would be greatly appreciated! I just can not find one...

**Reply by glenn20** [reply]
Yes, you can allocate the memory internally within your module, instead of passing in a bytearray. You just have to ensure that memory is protected from being garbage collected in the future by either ensuring some micropython object contains a pointer to that memory or adding it to the list of root pointers.

You will want to ensure you can allocate the memory from some fast memory pool on your device. I don't know how to do that, but believe it should be possible.

**Reply by glenn20** [reply]
Allocating the memory with `mp_obj_malloc()` probably won't be sufficient to solve your problem, since the same memory allocation method is used to allocate the memory using `buf = bytearray(size)`. If you are using a PSRAM device, you may want to allocate the buffer as soon as possible after device startup to maximise your chances of allocating from the limited faster RAM.

Oh - and the documentation for **all** the mp_ functions is the source code - there are a staggering number of mp_* functions and macros. However, I have found it quite productive just browsing code definitions.

**Reply by nspsck** [reply]
Thank you! Browsing definition is a really good way for me to learn these functions! And yes, you are right! `mp_obj_malloc()` is not going to solve my problem. 

However, I do not think that the allocated memory in the c module is not protected. The problem lies most likely in the fact, that I am allocating over half of the memory total. (250KB of 512KB) Which could leads to overwritting some memories used by micropython hosting repl. Allocating the memory is fine, working with it is also fine. Just, if I interrupt the device, or simply want to run the program the 2nd time, it forcely reboots. 

Also by running the gc module and doing some calculations, I came to the conclusion, that there should be enough memory available. Now I probably to narrow it done on how micropython interacts with the heap... However, gc module can be accessed in the c module, so am going to try it when I get back home. Hopefully that works...

**Reply by nspsck** [reply]
Update: gc.collect() or m_malloc() works, and the performance is as same as a bytearray.

**Reply by glenn20** [reply]
> However, I do not think that the allocated memory in the c module is not protected. The problem lies most likely in the fact, that I am allocating over half of the memory total. (250KB of 512KB) Which could leads to overwritting some memories used by micropython hosting repl. 

If you have allocated the memory correctly, it won't overwrite memory used by micropython **unless** you have not protected the allocated memory from garbage collection. If the memory does get claimed back by the gc before you are finished with it, then that memory will be re-used by micropython and anything could happen.

> Allocating the memory is fine, working with it is also fine. Just, if I interrupt the device, or simply want to run the program the 2nd time, it forcely reboots.

If I interpret what you said here as that you have soft-reset micropython and tried to run the program again, then one thing to watch out for:
- after soft reset, all the dynamic memory will be restored to the gc pool. However, if you keep a pointer to your allocated memory block in static memory, that will **NOT** be re-initialised and will continue to point to the address of the memory block from before the soft-reset - memory which will now be in use by other parts of the system. In fact, it is very difficult from a user module to check if your static data structures need to be reset because the user has soft-reset the device.

**Reply by nspsck** [reply]
You are right! Allocating the memory with `gc_collect()` and `m_malloc()` appearently are the right way to allocate the memory. And the allocated memory can be seen by the garbage collector from micropython correctly. Allocating memory with `malloc()` or `heap_caps_malloc()` (probably provided by esp-idf) is not the "correct" way, since the allocated memory can not be seen by the gc from micropython. _(This part is proven to be correct with some testing code in micropython)_ This proves that those 2 functions does not "talk" to micropython at all, hence there is no way for those to know, if the section of memory can be safely overwritten. (Maybe through freeRTOS, but I have 0 idea of it)

For the soft reset part. According to the [official documentation from micropython](https://docs.micropython.org/en/latest/library/machine.html?highlight=soft%20reset), you can see that it tries to keep REPL alive. So if that section where REPL uses is corrupted, this will probably trigger a hard reset. 
And, I believe, that the problem lies not in the "allocating" part, it actually lies in the "using" it part. My appologise for not presenting enough information at once. So the situation is: If I **USE** all of these 250KB allocated memory, I **can not** perform a re-run of the python script. If I **only uses roughly 120KB** of the allocated memory, **the re-runing works fine** and can be repeated as often as I wish. 
Further, I had a suspicion that micropython does not set alot RAM for the heap region. So I ran a little script as below:

```python
import gc

def heap():
  free = gc.mem_free() / 1024
  allocated = gc.mem_alloc() / 1024
  return free + allocated
  
def main():
    print("%dKB" % heap())

main()

```
According the [documentation](https://docs.micropython.org/en/v1.9.3/pyboard/library/gc.html), this should calculate the available heap in KB. And the result is as, somehow, expected... For an esp32 it returns roughly 104KB, for an esp32c3 it returns roughly 130KB. And for the esp32s3 n16r8 version, it returns only 4156KB which in translation (probably, perhaps, maybe): 4MB PSRAM + 60KB RAM. Since micropython run as a Task on the freeRTOS, I am not educated enough to assume anything about it's memory usage itself. But for the performance part, the problem reveals itself a little. **Micropython probably only uses PSRAM.** Again, this is purely speculative and I haven't found more facts to back it up, yet...

**Reply by iBigDane** [reply]
bytearray use RAM memory，not use PSRAM.You can test it with the following program.
run the program is fail
```
a = bytearray(200*1000)
```
run the program is success
```
a = bytearray(100*1000)
```
Because the esp32 spiram board leaves about 100KB of RAM

**Reply by nspsck** [reply]
Oh! That's cool! Hmm, what could be the cause of it then performing worse than if I allocate the memory directly using malloc or hesp_caps_malloc? Where lies the overhead?

**Reply by iBigDane** [reply]
https://github.com/micropython/micropython/blob/01c758e26a846eca42e9c4444639fa822adf005f/extmod/modframebuf.c#L226C1-L232C2
use the memset function,set one line pixel color.

**Reply by nspsck** [reply]
Actually, that's what I have been using (Or I think I am using it...). If you want to view the whole implementation you can check out [this](https://github.com/nspsck/lcd_binding_micropython/blob/master/lcd/driver/rm67162/rm67162.c#L353C1-L356C2).
```C
STATIC void fill_color_buffer(mp_lcd_rm67162_obj_t *self, uint16_t color, int len /*in pixel*/) {
    uint16_t *buffer = self->frame_buffer;
    size_t size = len;
    while (size--) {
        *buffer++ = color;
    }
    write_color(self, self->frame_buffer, len * 2);
}
```
This is probably, at least to my eyes, the same function as `memset()`. Meanwhile I found out the bug that causes mpy to reboot. That's because I set more memory in this function than I actually allocated. (Ye... This function, as same as `memset()`, does not check the end of the `frame_buffer`pointer, which should have been done.. Absolutely my bad.) If the way how `frame_buffer` is allocated has been changed the performance difference will occur. And I use [this](https://github.com/nspsck/lcd_binding_micropython/blob/master/lcd/driver/rm67162/rm67162.c#L73C1-L83C2) function to allocate memory for the buffer:
```C
STATIC void frame_buffer_alloc(mp_lcd_rm67162_obj_t *self, int len) {
    // create a constant DMA-enabled frambuffer.
    self->frame_buffer_size = len;
    //self->frame_buffer = heap_caps_malloc(self->frame_buffer_size, MALLOC_CAP_DMA);
    self->frame_buffer = gc_alloc(self->frame_buffer_size, 0);
    
    if (self->frame_buffer == NULL) {
        mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("Failed to allocate DMA'able framebuffer."));
    }
    memset(self->frame_buffer, 0, self->frame_buffer_size);
}
```
Using `gc_alloc()` results a lower performance than using `heap_caps_malloc()`.

**Reply by iBigDane** [reply]
For example, the screen resolution is 240*240, and the number of loops in your code is 57600. If you use memset to set the pixels of one row at a time, the number of loops is 240.I think so.
The gc_alloc() function should allocate SPIRAM

**Reply by nspsck** [reply]
Hi, thank you for all your advices so far. You are right! I checked later and `memset()` is indeed faster than my original function. It looks like the compiler knows how to optimise for 8-bits memory manipulation but does not know how to optimise for 16-bits memory manipulation. The source code of `memset()` seems to be this:
```C
void *memset(void *s, int c, size_t n) {
    if (c == 0 && ((uintptr_t)s & 3) == 0) {
        // aligned store of 0
        uint32_t *s32 = s;
        for (size_t i = n >> 2; i > 0; i--) {
            *s32++ = 0;
        }
        if (n & 2) {
            *((uint16_t*)s32) = 0;
            s32 = (uint32_t*)((uint16_t*)s32 + 1);
        }
        if (n & 1) {
            *((uint8_t*)s32) = 0;
        }
    } else {
        uint8_t *s2 = s;
        for (; n > 0; n--) {
            *s2++ = c;
        }
    }
    return s;
}
```
Sadly, If i just call that function as below
```C
memset(self->frame_buffer, color, len * 2); 
// the color is 2 bytes, and since color != 0, it will be a byte-wise operation
// hence if I were calling len, only half of the screen area will be drawn.
```
The color on the display seems to be broken. I.e. (255, 0, 0) will look like pink/purple and (0, 255, 0) will be green and (0, 0, 255) will just stay black. So I tried to modify my own code in order to match the performance. So I did this:
```C
STATIC void fill_color_buffer(mp_lcd_rm67162_obj_t *self, uint16_t color, int len /*in pixel*/) {
    uint32_t *buffer = (uint32_t *)self->frame_buffer;
    // this ensures that the framebuffer is overfilled rather than unfilled.
    size_t size = (len + 1) / 2; 
    while (size--) {
        // ye, well, this should not work, but it works................but why?
        *buffer++ = color;
    }
    write_color(self, self->frame_buffer, len * 2);
}
```
and I've noticed, the `*buffer++ = color;` part should not work since the upper 16 bits is not controled by me at all... but to my surprise.. it works flawlessly...... Could you please give some possible explanations? I am using a display controler named RM67162. 
Either way, the performance are very similar right now.

Also, the bytearray and gc_alloc thing... Due to this issue https://github.com/micropython/micropython/issues/11853 , I can not make any accurate statements about what `gc` actually allocates.

**Reply by iBigDane** [reply]
This is my mistake, if you are using rgb565 color mode, the method of setting pixels cannot use memset().
https://github.com/micropython/micropython/blob/01c758e26a846eca42e9c4444639fa822adf005f/extmod/modframebuf.c#L130C1-L138C2
As for the performance problem, there is no idea at present.
